Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,162 @@
-    } else {
-        for (unsigned i = 0; i < count; i++) {
-            _locations[i] = (float)i / (float)(count - 1);
-        }
+    __CGGradient(CGColorSpaceRef colorSpace, CFArrayRef components, const CGFloat* locations)
+        : _colorSpace(colorSpace),
+          _colorComponents(CopyColorComponents(colorSpace, components)),
+          _stopLocations(CopyLocations(locations, CFArrayGetCount(components))) {
     }
 
-    _count = count;
-}
+    inline CGColorSpaceRef GetColorSpace() const {
+        return _colorSpace;
+    }
 
-void __CGGradient::initWithColors(CFArrayRef componentsArr, const float* locations, CGColorSpaceRef colorspace) {
-    NSArray* components = (NSArray*)componentsArr;
-    size_t componentCount = 0;
-    __CGColorSpace* cs = (__CGColorSpace*)colorspace;
-
-    _colorSpaceModel = cs->colorSpaceModel;
-
-    switch (_colorSpaceModel) {
-        case kCGColorSpaceModelRGB:
-            componentCount = 4;
-            _format = _ColorABGR;
-            break;
-        case kCGColorSpaceModelMonochrome:
-            componentCount = 2;
-            _format = _ColorGrayscale;
-            break;
-        default:
-            UNIMPLEMENTED_WITH_MSG("Unsupported colorspace used to create gradiant.");
-            break;
+    inline const std::vector<CGFloat>& GetColorComponents() const {
+        return _colorComponents;
     }
 
-    int count = [components count];
+    inline const std::vector<CGFloat>& GetStopLocations() const {
+        return _stopLocations;
+    }
 
-    _components = new CGFloat[count * componentCount];
+    woc::StrongCF<CGColorSpaceRef> _colorSpace;
+    std::vector<CGFloat> _colorComponents;
+    std::vector<CGFloat> _stopLocations;
 
-    for (int i = 0; i < count; i++) {
-        UIColor* curColor = [components objectAtIndex:i];
+private:
+    std::vector<CGFloat> CopyLocations(const CGFloat* locations, size_t count) {
+        // locations can be null
+        std::vector<CGFloat> result;
+        RETURN_RESULT_IF(count == 0, result);
 
-        if (curColor) {
-            const __CGColorQuad* color = [curColor _getColors];
-            memcpy(&_components[i * componentCount], color, sizeof(CGFloat) * componentCount);
+        if (locations) {
+            result = std::vector<CGFloat>(locations, locations + count);
         } else {
-            memset(&_components[i * componentCount], 0, sizeof(CGFloat) * componentCount);
+            result.reserve(count);
+
+            for (size_t i = 0; i < count; i++) {
+                result.push_back((CGFloat)i / (CGFloat)(count));
+            }
         }
-    }
 
-    _locations = new float[count];
+        return result;
+    }
 
-    if (locations) {
-        memcpy(_locations, locations, sizeof(float) * count);
-    } else {
-        for (int i = 0; i < count; i++) {
-            _locations[i] = (float)i / (float)(count - 1);
+    std::vector<CGFloat> CopyColorComponents(CGColorSpaceRef colorspace, CFArrayRef components) {
```

Comment:
LOL yes, but
it's because we want to do the cool constructor initialization of prams, also i went back and forth on this with dustin.
leaving it as it is.

---
In reply to: [107501560](https://github.com/Microsoft/WinObjC/pull/2249#discussion_r107501560) [](ancestors = 107501560)